symfony/console/3.3/bin/console

```diff
@@ -3,18 +3,29 @@
 
 use App\Kernel;
 use Symfony\Bundle\FrameworkBundle\Console\Application;
+use Symfony\Component\Console\Input\ArgvInput;
 use Symfony\Component\Debug\Debug;
 
 set_time_limit(0);
 
 require dirname(__DIR__).'/vendor/autoload.php';
 
 if (!class_exists(Application::class)) {
-    throw new \RuntimeException('You need to add "symfony/framework-bundle" as a Composer dependency.');
+    throw new RuntimeException('You need to add "symfony/framework-bundle" as a Composer dependency.');
 }
 
-Kernel::bootstrapCli($_SERVER['argv']);
-Kernel::bootstrapEnv();
+$input = new ArgvInput();
+if (null !== $_ENV['APP_ENV'] = $input->getParameterOption(['--env', '-e'], null, true)) {
+    putenv('APP_ENV='.$_ENV['APP_ENV']);
+    // force loading .env files when --env is defined
+    $_SERVER['APP_ENV'] = null;
```

Comment:
See previous comment from @weaverryan: we want .env files to be loaded also when --env is used, even for prod.
There is only one way to opt-out from .env files, which is to set a real APP_ENV var to prod.

Reply:
I don't see the use-case for this to add such an inconsistency. People will run bin/console --env=prod locally which might work. Then they run exactly the same command when deployed, it it might not work because it tries to load env files but the DotEnv is not installed. Thus they get an exception.

Reply:
that's the opposite: Dotenv is *not* loaded in prod, when APP_ENV is set.

Reply:
People that do not define a real APP_ENV but run commands using --env=prod (like sf 3 style) will think it works because it works locally. But it will not work when deployed without the DotEnv.

Reply:
https://github.com/symfony/skeleton/pull/122 should address the concern.

Reply:
if people defined APP_ENV on their prod, they opt-ed in for using real env vars.
Then it's their job to make the process robust.

Reply:
So there are two camps:
1. people not using real env vars who want to use DotEnv also for prod
    - they just need to make DotEnv a real dependency
    - and remove `if ('prod' !== $_SERVER['APP_ENV']) {` from bootstrap
2. people using real env vars
    - `if ('prod' !== $_SERVER['APP_ENV']) {` and `require-dev=DotEnv` secures them against 
      - forgetting to define env vars and accidentally loading .env without realizing it
      - defining APP_ENV but some dependency moved DotEnv to require without realizing it, and thus loading .env params accidentally

symfony/skeleton#122 will break the safeguard for 2)
And `$_SERVER['APP_ENV'] = null;` makes the bahavior inconsistent for 1)

Reply:
I'm not sure I understand what you're proposing to change?

Reply:
What I'm proposing is not do symfony/skeleton#122 and remove `$_SERVER['APP_ENV'] = null;` so that `APP_ENV=prod bin/console` and `bin/console --env=prod` behaves the same.

>  A fatal error is no better - that didn't save anything really. Only proper deployment process can fail proof a prod. 

Any decent deployment will test if the app runs correctly before switching it live. So yes, throwing an exception when env vars are missing is exactly part of the deployment process you describe. If you remove that behavior, only complicated functional tests can cover loading the wrong environment.

Reply:
> remove $_SERVER['APP_ENV'] = null; so that APP_ENV=prod bin/console and bin/console --env=prod behaves the same.

I disagree with this part - this would make things more confusing. Passing the env flag changes the environment only, and then all the normal rules should be followed. Loading the env vars for -env=dev but suddenly not for -env=prod is weird behavior.

Reply:
To me --env=... is just an alias for APP_ENV=... 
This is what people wanted in https://github.com/symfony/symfony/issues/28984 because it's sometimes hard to set env variables. If those two things are not doing the same thing, it's confusing and can cause unexpected behavior in my opinions. But I guess it depends on the perspective.

Reply:
When first submitting the PR it behaved as you describe Tobion, but now I understand the rationale of what Ryan's perspective and it makes sense to me.

Reply:
In case one is wondering, Laravel loads .env.prod when --env=prod is used ;)

Reply:
I still don't see the use-case. If I use real env vars, why would I use .env.prod at all? If I don't use real env, loading .env.prod for--env=prod is not gonna work when the app is deployed by default.
But in the end, as long as people understand the differences between --env and APP_ENV, this is not a problem but more a doc issue.